iceoryx_examples/waitset/README.md

```diff
@@ -1,6 +1,586 @@
+    handled differently.
+
+ 4. `ice_waitset_sync`: We use the WaitSet to trigger a cyclic call which should 
+    execute an algorithm every 100ms.
+
+ 5. `ice_waitset_trigger`: We create our own class which can be attached to a
+    WaitSet and trigger events.
+
+## Examples
+
+All our examples require a running `iox-roudi` and a running `iox-ex-waitset-publisher`
+which is sending the data. The publisher does not contain any _WaitSet_ specific
+logic and is explained in detail in the [icedelivery example](../icedelivery/).
+
+### Gateway
+
+We have a list of subscribers which can be subscribed to any arbitrary topic
+and everytime we received a sample we would like to send the bytestream to a socket,
+write it into a file or print it to the console. But whatever we choose to do
+we perform the same task for all the subscribers.
+
+Lets start by implementing our callback which prints the subscriber pointer, the
+payload size and the payload pointer to the console.
+```cpp
+void subscriberCallback(iox::popo::UntypedSubscriber* const subscriber)
+{
+    subscriber->take().and_then([&](iox::popo::Sample<const void>& sample) {
+        std::cout << "subscriber: " << std::hex << subscriber << " length: " << std::dec
+                  << sample.getHeader()->m_info.m_payloadSize << " ptr: " << std::hex << sample.getHeader()->payload()
+                  << std::endl;
+    });
+}
+```
+A _Trigger_ always requires a callback which has the following signature 
+`void (TriggerOrigin)`. In our example the _TriggerOrigin_ is a
+`iox::popo::UntypedSubscriber*` which we use to acquire the latest sample which the subscriber
+has received (`take()`). When `take()` was successful we print our message to
+the console inside of the `and_then` lambda.
+
+In our `main` function we create a _WaitSet_ after we registered at our central
+broker RouDi and attach our `shutdownTrigger` to it to handle `CTRL+c` events. 
+```cpp
+iox::popo::WaitSet waitset;
+
+shutdownTrigger.attachTo(waitset);
+```
+
+After that we create a vector of 2 subscribers, subscribe and attach them to a
+_WaitSet_ with the event `HAS_NEW_SAMPLES` and the `subscriberCallback`. Everytime one 
+of the subscribers is receiving a new sample it will trigger the _WaitSet_.
+```cpp
+iox::cxx::vector<iox::popo::UntypedSubscriber, 2> subscriberVector;
+for (auto i = 0; i < subscriberVector.capacity(); ++i)
+{
+    subscriberVector.emplace_back(iox::capro::ServiceDescription{"Radar", "FrontLeft", "Counter"});
+    auto& subscriber = subscriberVector.back();
+
+    subscriber.subscribe();
+    subscriber.attachTo(waitset, iox::popo::SubscriberEvent::HAS_NEW_SAMPLES, 1, subscriberCallback);
+}
+```
+
+Now our system is prepared and ready to work. We enter the event loop which 
+starts with a call to our _WaitSet_ (`waitset.wait()`). This call will block until
+one or more events triggered the _WaitSet_. After the call returned we get a
+vector filled with all the _Triggers_ which were triggered.
```

Comment:
It could happen but I would say it is a very rare case. One example where this could happen would be that a subscriber is attached to a WaitSet and that this subscriber triggers the WaitSet since it received a new sample. 

Meanwhile in a first thread we called `WaitSet.wait()` whilst in a second thread someone, exactly in the moment after the arrival of the sample, acquires and release the sample. When the WaitSet then iterates over the vector of triggers it would seem like no trigger was triggered.

But it would just mean that the user would receive an empty vector. I think this wouldn't be a problem. 

@budrus @MatthiasKillat  do you think this is a problem? I can fix it easily.

Reply:
If the user does not change the trigger state in parallel, this should not happen. If they do, they have to live with the consequence